File: app/src/main/java/com/github/nikololoshka/pepegaschedule/schedule/pair/DateRange.java
package com.github.nikololoshka.pepegaschedule.schedule.pair;

import android.os.Parcel;

import androidx.annotation.Nullable;

import com.github.nikololoshka.pepegaschedule.schedule.pair.exceptions.InvalidDateException;
import com.github.nikololoshka.pepegaschedule.schedule.pair.exceptions.InvalidFrequencyForDateException;
import com.github.nikololoshka.pepegaschedule.schedule.pair.exceptions.InvalidIntersectDateException;
import com.github.nikololoshka.pepegaschedule.schedule.pair.exceptions.InvalidPairParseException;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.Locale;
import java.util.Objects;

public class DateRange extends DateItem {

    @Nullable
    private Calendar mDateStart;
    @Nullable
    private Calendar mDateEnd;

    private FrequencyEnum mFrequency;

    DateRange() {
        mDateStart = null;
        mDateEnd = null;
        mFrequency = FrequencyEnum.EVERY;
    }

    private DateRange(Parcel in) {
        mDateStart = (Calendar) in.readSerializable();
        mDateEnd = (Calendar) in.readSerializable();
        mFrequency = (FrequencyEnum) in.readSerializable();
    }

    public DateRange(DateRange dateRange) {
        mDateStart = (Calendar) (dateRange.mDateStart != null ? dateRange.mDateStart.clone() : null);
        mDateEnd = (Calendar) (dateRange.mDateEnd != null ? dateRange.mDateEnd.clone() : null);
        mFrequency = dateRange.mFrequency;
    }

    public static DateRange of(String start, String end, FrequencyEnum frequency) {
        DateFormat format = new SimpleDateFormat("dd.MM.yyyy", Locale.ROOT);
        format.setLenient(false);
        Calendar dateStart = new GregorianCalendar();
        Calendar dateEnd = new GregorianCalendar();
        try {
            dateStart.setTime(format.parse(start));
            dateEnd.setTime(format.parse(end));
        } catch (ParseException ignored) {
            throw new InvalidDateException(String.format("Start: %s; End: %s; Frequency: %s",
                    start, end, frequency.tag()));
        }

        // check day of week
        DayOfWeek.valueOf(dateStart);
        DayOfWeek.valueOf(dateEnd);

        long diffInMillis = dateEnd.getTimeInMillis() - dateStart.getTimeInMillis();
        if ((diffInMillis / 86400000) % frequency.period() != 0) {
            throw new InvalidFrequencyForDateException(String.format("Start: %s; End: %s; Frequency: %s",
                    start, end, frequency.tag()));
        }

        DateRange dateRange = new DateRange();
        dateRange.mDateStart = dateStart;
        dateRange.mDateEnd = dateEnd;
        dateRange.mFrequency = frequency;
        return dateRange;
    }

    public static final Creator<DateRange> CREATOR = new Creator<DateRange>() {
        @Override
        public DateRange createFromParcel(Parcel in) {
            return new DateRange(in);
        }

        @Override
        public DateRange[] newArray(int size) {
            return new DateRange[size];
        }
    };

    @Nullable
    public Calendar dateStart() {
        return mDateStart;
    }

    @Nullable
    public Calendar dateEnd() {
        return mDateEnd;
    }

    @Override
    public FrequencyEnum frequency() {
        return mFrequency;
    }

    @Override
    public DayOfWeek dayOfWeek() {
        return DayOfWeek.valueOf(mDateStart);
    }

    @Override
    public String compactDate() {
        if (mDateStart == null || mDateEnd == null) {
            return "";
        }

        SimpleDateFormat dateFormat = new SimpleDateFormat("dd.MM", Locale.ROOT);
        return dateFormat.format(mDateStart.getTime()) + "-"
                + dateFormat.format(mDateEnd.getTime()) + " " + mFrequency.text();
    }

    @Override
    public String fullDate() {
        if (mDateStart == null || mDateEnd == null) {
            return "";
        }

        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy.MM.dd", Locale.ROOT);
        return dateFormat.format(mDateStart.getTime()) + "-"
                + dateFormat.format(mDateEnd.getTime());
    }

    @Override
    public Calendar minDate() {
        return mDateStart;
    }

    @Override
    public Calendar maxDate() {
        return mDateEnd;
    }

    @Override
    public void load(JSONObject loadObject) throws JSONException {
        String frequency = loadObject.getString("frequency");
        String date = loadObject.getString("date");

        if (frequency.equals(FrequencyEnum.EVERY.tag())) {
            mFrequency = FrequencyEnum.EVERY;
        } else if (frequency.equals(FrequencyEnum.THROUGHOUT.tag())) {
            mFrequency = FrequencyEnum.THROUGHOUT;
        } else {
            throw new InvalidPairParseException("Invalid frequency: " + frequency);
        }

        String[] dates = date.split("-");
        if (dates.length != 2) {
            throw new InvalidDateException(date);
        }

        String startDate = dates[0];
        String endDate = dates[1];

        DateFormat format = new SimpleDateFormat("yyyy.MM.dd", Locale.ROOT);
        format.setLenient(false);
        Calendar dateStart = new GregorianCalendar();
        Calendar dateEnd = new GregorianCalendar();

        try {
            dateStart.setTime(format.parse(startDate));
            dateEnd.setTime(format.parse(endDate));
        } catch (ParseException ignored) {
            throw new InvalidDateException(String.format("Start: %s; End: %s; Frequency: %s",
                    startDate, endDate, mFrequency.tag()));
        }

        long diffInMillis = dateEnd.getTimeInMillis() - dateStart.getTimeInMillis();
        if ((diffInMillis / 86400000) % mFrequency.period() != 0) {
            throw new InvalidFrequencyForDateException(String.format("Start: %s; End: %s; Frequency: %s",
                    startDate, endDate, mFrequency.tag()));
        }

        mDateStart = dateStart;
        mDateEnd = dateEnd;
    }

    @Override
    public boolean intersect(DateItem dateItem) {
        if (mDateStart != null) {
            if (dateItem instanceof DateSingle) {
                Calendar dateSingle = ((DateSingle) dateItem).date();
                Calendar date = (Calendar) mDateStart.clone();

                while (date.compareTo(mDateEnd) <= 0) {
                    if (Objects.equals(date, dateSingle)) {
                        return true;
                    }
                    date.add(Calendar.DAY_OF_MONTH, mFrequency.period());
                }
                return false;
            }
            if (dateItem instanceof DateRange) {
                DateRange dateRange = (DateRange) dateItem;

                if (dateRange.dateStart() == null) {
                    throw new InvalidIntersectDateException(String.format("Can not intersect date: '%s' and '%s'",
                            this.toString(), dateItem.toString()));
                }

                Calendar firstDate = (Calendar) dateRange.dateStart().clone();
                Calendar secondDate = (Calendar) mDateStart.clone();

                int i = 0;
                while (firstDate.compareTo(dateRange.dateEnd()) <= 0) {
                    secondDate.add(Calendar.DAY_OF_MONTH, i * mFrequency.period());
                    i = 0;
                    while (secondDate.compareTo(mDateEnd) <= 0) {
                        if (firstDate.equals(secondDate)) {
                            return true;
                        }
                        secondDate.add(Calendar.DAY_OF_MONTH, mFrequency.period());
                        i--;
                    }
                    firstDate.add(Calendar.DAY_OF_MONTH, dateRange.frequency().period());
                }
                return false;
            }
        }
        throw new InvalidIntersectDateException(String.format("Can not intersect date: '%s' and '%s'",
                this.toString(), dateItem.toString()));
    }

    @Override
    public void save(JSONArray saveArray) throws JSONException {
        JSONObject dateObject = new JSONObject();
        dateObject.put("frequency", mFrequency.tag());
        dateObject.put("date", fullDate());
        saveArray.put(dateObject);
    }

    @Override
    public int compareTo(DateItem o) {
        if (o instanceof DateSingle) {
            DateSingle dateSingle = (DateSingle) o;
            return fullDate().compareTo(dateSingle.fullDate());
        }
        if (o instanceof DateRange) {
            DateRange dateRange = (DateRange) o;
            return fullDate().compareTo(dateRange.fullDate());
        }
        throw new IllegalArgumentException(String.format("Not compare objects: '%s' and '%s'",
                this.toString(), o.toString()));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        DateRange dateRange = (DateRange) o;
        return Objects.equals(mDateStart, dateRange.mDateStart) &&
                Objects.equals(mDateEnd, dateRange.mDateEnd) &&
                mFrequency == dateRange.mFrequency;
    }

    @Override
    public int hashCode() {
        return Objects.hash(mDateStart, mDateEnd, mFrequency);
    }

    @Override
    public String toString() {
        if (mDateStart == null || mDateEnd == null) {
            return "null - null null";
        }

        SimpleDateFormat dateFormat = new SimpleDateFormat("dd.MM.yyyy", Locale.ROOT);
        return String.format("%s - %s %s", dateFormat.format(mDateStart.getTime()),
                dateFormat.format(mDateEnd.getTime()), mFrequency.text());
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeSerializable(mDateStart);
        dest.writeSerializable(mDateEnd);
        dest.writeSerializable(mFrequency);
    }
}

File: app/src/main/java/com/vereshchagin/nikolay/stankinschedule/model/schedule/pair/DateRange.kt
package com.vereshchagin.nikolay.stankinschedule.model.schedule.pair

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import com.google.gson.JsonObject
import com.vereshchagin.nikolay.stankinschedule.R
import com.vereshchagin.nikolay.stankinschedule.utils.DateUtils
import org.joda.time.DateTime
import org.joda.time.Days
import org.joda.time.LocalDate
import org.joda.time.format.DateTimeFormat

class DateRange : DateItem {

    internal val start: DateTime
    internal val end: DateTime
    private val frequency: Frequency
    private lateinit var dayOfWeek: DayOfWeek

    constructor(parcel: Parcel) {
        start = parcel.readSerializable() as DateTime
        end = parcel.readSerializable() as DateTime
        frequency = parcel.readSerializable() as Frequency

        init()
    }

    constructor(firstText: String, secondText: String, frequencyDate: Frequency,
                pattern: String = JSON_DATE_PATTERN) {
        try {
            val formatter = DateTimeFormat.forPattern(pattern)
            start = formatter.parseDateTime(firstText)
            end = formatter.parseDateTime(secondText)
            frequency = frequencyDate

        } catch (e: Exception) {
            throw DateParseException(
                "Invalid parse date: $firstText and $secondText",
                "$firstText - $secondText",
                e
            )
        }

        init()
    }

    constructor(text: String, frequencyDate: Frequency, pattern: String = JSON_DATE_PATTERN) {
        val dates = text.split('-')
        if (dates.size != 2) {
            throw DateParseException(
                "Invalid date text: $text, $dates, frequency: $frequencyDate",
                text
            )
        }

        val (firstText, secondText) = dates
        try {
            val formatter = DateTimeFormat.forPattern(pattern)
            start = formatter.parseDateTime(firstText)
            end = formatter.parseDateTime(secondText)
            frequency = frequencyDate

        } catch (e: Exception) {
            throw DateParseException(
                "Invalid parse date: $firstText and $secondText",
                "$firstText - $secondText",
                e
            )
        }

        init()
    }

    constructor(firstDate: DateTime, secondDate: DateTime, frequencyDate: Frequency) {
        start = firstDate
        end = secondDate
        frequency = frequencyDate

        init()
    }

    constructor(obj: JsonObject) : this (
        obj[JSON_DATE].asString,
        Frequency.of(obj[JSON_FREQUENCY].asString)
    )

    private fun init() {
        if (DayOfWeek.of(start) != DayOfWeek.of(end)) {
            throw DateDayOfWeekException(
                "Invalid day of week: $start - $end"
            )
        }
        dayOfWeek = DayOfWeek.of(start)

        val days = Days.daysBetween(start, end).days
        if (days == 0 || days % frequency.period != 0) {
            throw DateFrequencyException(
                "Invalid frequency: $start - $end, ${frequency.tag}",
                this.toString(), frequency
            )
        }
    }

    override fun dayOfWeek(): DayOfWeek = dayOfWeek

    override fun frequency(): Frequency = frequency

    override fun toJson(): JsonObject {
        return JsonObject().apply {
            addProperty(
                JSON_DATE,
                start.toString(JSON_DATE_PATTERN) + "-" + end.toString(JSON_DATE_PATTERN)
            )
            addProperty(JSON_FREQUENCY, frequency.tag)
        }
    }

    override fun intersect(item: DateItem): Boolean {
        if (item is DateSingle) {
            var it = start
            while (it.isBefore(end) || it == end) {
                if (it == item.date) {
                    return true
                }
                it = it.plusDays(frequency.period)
            }
            return false
        }

        if (item is DateRange) {
            var firstIt = start
            var secondIt = item.start

            while (firstIt.isBefore(end) || firstIt == end) {
                while (secondIt.isBefore(item.end) || secondIt == item.end) {
                    if (firstIt == secondIt) {
                        return true
                    }
                    secondIt = secondIt.plusDays(item.frequency.period)
                }
                firstIt = firstIt.plusDays(frequency.period)
            }

            return false
        }

        throw IllegalArgumentException("Invalid intersect object: $item")
    }

    override fun isBefore(item: DateItem): Boolean {
        if (item is DateSingle) {
            return start.isBefore(item.date) && end.isBefore(item.date)
        }

        if (item is DateRange) {
            return start.isBefore(item.start) && end.isBefore(item.end)
        }

        throw IllegalArgumentException("Invalid compare object: $item")
    }

    override fun startDate(): LocalDate {
        return start.toLocalDate()
    }

    override fun endDate(): LocalDate {
        return end.toLocalDate()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as DateRange

        if (start != other.start) return false
        if (end != other.end) return false
        if (frequency != other.frequency) return false
        if (dayOfWeek != other.dayOfWeek) return false

        return true
    }

    override fun hashCode(): Int {
        var result = start.hashCode()
        result = 31 * result + end.hashCode()
        result = 31 * result + frequency.hashCode()
        result = 31 * result + dayOfWeek.hashCode()
        return result
    }

    override fun toString(): String {
        return start.toString(DateUtils.PRETTY_FORMAT) +
            "-" +  end.toString(DateUtils.PRETTY_FORMAT)
    }

    override fun toString(context: Context): String {
        return this.toString() + context.resources.getStringArray(
                R.array.frequency_simple_list
            )[
            when(frequency) {
                Frequency.EVERY -> 0
                Frequency.THROUGHOUT -> 1
                else -> throw RuntimeException("Unknown frequency: $frequency")
            }
        ]
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeSerializable(start)
        parcel.writeSerializable(end)
        parcel.writeSerializable(frequency)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object {
        @JvmField
        val CREATOR = object: Parcelable.Creator<DateRange> {
            override fun createFromParcel(parcel: Parcel): DateRange {
                return DateRange(parcel)
            }

            override fun newArray(size: Int): Array<DateRange?> {
                return arrayOfNulls(size)
            }
        }
    }
}